// kit/common/index.js
'use strict';

const ObjectCommon = require('./object');
const classInfo = require('./class-info');
const ArrayCommon = require('./array');
const FunctionsCommon = require('./functions');
const StringCommon = require('./string');

module.exports = Object.assign({
  clearTags: require('./clear-tags'),
  datetime: require('./datetime'),
  encodeUnicode: require('./encode-unicode'),
  filesize: require('./filesize'),
  shortenByWords: require('./shorten-by-words'),
  textToHtml: require('./text-to-html'),
  trimFilename: require('./trim-filename'),
  createFunction: require('./create-function'),
  expression: require('./expression'),
  interpolate: require('./interpolate'),
  round: require('./round'),
  argsList: require('./args-list'),
  unicode: require('./unicode'),
  promiseFactory: require('./promise-factory'),
  mapToObject: require('./map-to-object'),
  objectsToArray: require('./objects-to-array'),
  methodsNames: require('./methods-names'),
  injectMethods: require('./inject-methods'),
  selfMethod: require('./self-method'),
  trim: require('./trim'),
  orUndefined: require('./or-undefined'),
  random: require('./random'),
  range: require('./range'),
  arrayToTrueObject: require('./array-to-true-object'),
  trueObjectToArray: require('./true-object-to-array'),
  uniqueId: require('./unique-id'),
  uid: require('./unique-id'),
  query: require('./query'),
  tmpl: require('./tmpl'),
  classInfo,

  EventEmitter: require('./event-emitter'),

  resolver, isObject, notObject,

},
classInfo.getMethods(ObjectCommon),
classInfo.getMethods(ArrayCommon),
classInfo.getMethods(FunctionsCommon),
classInfo.getMethods(StringCommon)
);

function resolver(dir) {
  return file => path.resolve(dir, file);
}

function isObject(item) {
  return item && typeof item === 'object' && !Array.isArray(item);
}

function notObject(item) {
  return isObject(item);
}
