// plp/bg/manager.class.js
'use strict';

const $plp = {

};

const state = {

};

class BackgroundManager {

  constructor (elem) {
    var self = this;
    self.elem  = elem;
    self.images = [];
    self.current = -1;
  }

  setImages (urls) {
    var self = this;
    self.images = urls;
    self.restore();
  }

  random () {
    var self = this;
    self.current = Math.floor(Math.random() * self.images.length);
    self.setIndex(self.current);
  }

  next () {
    var self = this;
    self.current++;
    if (self.current >= self.images.length) {
      self.current = 0;
    }
    self.setIndex(self.current);
  }

  prev () {
    var self = this;
    self.current--;
    if (self.current < 0) {
      self.current = self.images.length - 1;
    }
    self.setIndex(self.current);
  }

  setIndex (i) {
    var self = this;
    self.elem.style.background = `url('${self.images[self.current]}')`;
    self.save();
  }

  save () {
    var self = this;
    if (self === -1) return;
    localStorage.setItem('libro-bg-index', `${self.current}`);
  }

  restore () {
    var self = this,
      i = +localStorage.getItem('libro-bg-index');
    if (!self.images[i]) return;
    self.current = i;
    self.setIndex(i);
  }

}

module.exports = BackgroundManager;
