// plp/io/file-slice.class.js
'use strict';

const _ = {

};

const state = {

};

class FileSlice {

  constructor (blob, begin, end) {
    var self = this;
    self.blob = blob;
    self.begin = begin;
    self.end = end;
  }

  get size () {
    var self = this;
    return self.blob.size;
  }

  readAs (type) {
    var self = this,
      blob = self.blob;

    type = type[0].toUpperCase() + type.slice(1);

    return new Promise((resolve, reject) => {
      var reader = new FileReader();
      reader.onload = function (ev) {
        resolve(ev.target.result);
      };
      reader[`readAs${type}`](blob);
    });
  }

  toDataUrl () {
    var self = this;
    return self.readAs('dataURL');
  }
  
  dataUrl () {
    var self = this;
    return self.toDataUrl();
  }

  toBuffer () {
    var self = this;
    return self.readAs('arrayBuffer');
  }
  
  buffer () {
    var self = this;
    return self.toBuffer();
  }

  toBinaryString () {
    var self = this;
    return self.readAs('binaryString');
  }
  
  binaryString () {
    var self = this;
    return self.toBinaryString();
  }

  toText () {
    var self = this;
    return self.readAs('text');
  }
  
  text () {
    var self = this;
    return self.toText();
  }

}

module.exports = FileSlice;
