// kit/client/proto/tags.js
var tags = {
  a: true,
  abbr: true,
  address: true,
  area: true,
  article: true,
  aside: true,
  audio: true,
  b: false,
  base: false,
  bdi: false,
  bdo: false,
  blockquote: true,
  body: true,
  br: false,
  button: false,
  canvas: true,
  caption: true,
  cite: true,
  code: false,
  col: true,
  colgroup: true,
  datalist: false,
  dd: false,
  del: false,
  details: true,
  dfn: false,
  dialog: true,
  div: true,
  dl: false,
  dt: false,
  em: false,
  embed: true,
  fieldset: true,
  figcaption: true,
  figure: true,
  footer: true,
  form: true,
  h1: true,
  h2: true,
  h3: true,
  h4: true,
  h5: true,
  h6: true,
  head: true,
  header: true,
  hr: false,
  i: false,
  iframe: false,
  img: false,
  input: false,
  ins: false,
  kbd: false,
  keygen: false,
  label: false,
  legend: true,
  li: true,
  link: false,
  main: true,
  map: true,
  mark: true,
  menu: true,
  menuitem: true,
  meta: false,
  meter: false,
  nav: true,
  noscript: false,
  object: false,
  ol: true,
  optgroup: false,
  option: true,
  output: true,
  p: true,
  param: false,
  pre: true,
  progress: false,
  q: false,
  rp: false,
  rt: false,
  ruby: false,
  s: false,
  samp: false,
  script: false,
  section: true,
  select: false,
  small: false,
  source: true,
  span: false,
  strong: false,
  style: true,
  sub: false,
  summary: true,
  sup: false,
  table: true,
  tbody: true,
  td: true,
  textarea: false,
  tfoot: true,
  th: true,
  thead: true,
  time: false,
  title: true,
  tr: true,
  track: false,
  u: true,
  ul: true,
  video: true,
  wbr: false,
  svg: true,
};

module.exports = tags;
