// plp/path.js
'use strict';

const _ = {
  extend: require('./extend'),
  trim: require('./trim'),
};

const state = {
  splitter: '/',
};

function path() {

}

_.extend(path, {
  join,
  resolve,
  basename,
  extname,
  split,
});

function basename(p, ext) {
  var self = this,
    items = split(p),
    name = items.pop();
  return ext ? _.trim.right(name, ext) : name;
}

function extname(p) {
  var self = this,
    name = basename(p);
  if (!name.includes('.'))
    return '';
  return '.' + name.split('.').pop();
}

function split(p) {
  return p.split(state.splitter);
}

function resolve(...paths) {
  var p = join(...paths),
    items = split(p);
  while (items.indexOf('..') > -1) {
    let index = items.indexOf('..');
    items.splice(index - 1, 2);
  }
  return join(...items);
}

function join(...paths) {
  var double = state.splitter + state.splitter,
    p = paths.join(state.splitter);
  while (p.indexOf(double) > -1)
    p = p.replace(double, state.splitter);
  return p;
}

module.exports = path;
