// plp/query.js
'use strict';

const _ = {
  extend: require('./extend'),
  flatten: require('./flatten'),
  unflatten: require('./unflatten'),
};

function query(...args) {
  return typeof args[0] === 'string'
    ? parse(...args)
    : stringify(...args);
}

function parse(s, spl = '&') {
  var str = s.split('?').pop(),
    pairs = str.split(spl).map(s => s.trim().split('=')),
    data = {};
  if (!s) return {};
  pairs.forEach(pair => {
    var n = pair[0],
      name = n.endsWith('[]') ? n.slice(0, -2) : n,
      val = pair[1];
    if (data[name] && !Array.isArray(data[name])) {
      data[name] = [data[name]];
    }
    if (Array.isArray(data[name]) && n.endsWith('[]')) {
      return data[name].push(JSON.parse(decodeURIComponent(val)));
    }
    val = decodeURIComponent(val);
    try {
      val = JSON.parse(val);
    } catch(err) {}
    data[name] = val;
  });
  return _.unflatten(data);
}

function stringify(d, spl = '&') {
  var data = _.flatten(d, undefined, true);
  return Object
    .keys(data)
    .map(key => {
      var val = data[key];
      if (Array.isArray(val)) {
        return val
          .map(v => `${key}[]=${encodeURIComponent(v)}`)
          .join(spl);
      }
      return `${key}=${encodeURIComponent(val)}`;
    })
    .join(spl);
}

_.extend(query, {
  parse, stringify,
});

module.exports = query;
