// plp/legacy/local.js
class Local {

  constructor () {
    var self = this;
    self.path = '/';
    self.domain = location.hostname;
    self.days = 7;
  }

  // Cookies ===================================================================

  get cookies () {
    var self = this;
    return self.getCookies();
  }

  getCookieDate (d) {
    var self = this,
      days = d || self.days,
      date = new Date(Date.now() + days * 24 * 60 * 60 * 1000);
    return date.toUTCString();
  }

  getOptionsString (days) {
    var self = this;
    return `expires=${self.getCookieDate(days)}; path=${self.path}; domain=${self.domain};`;
  }

  getCookieString (name, value, days) {
    var self = this;
    return `${name}=${encodeURIComponent(value)}; ${self.getOptionsString(days)}`;
  }

  setCookie (name, value, days) {
    var self = this;
    document.cookie = self.getCookieString(name, value, days);
  }

  removeCookie (name) {
    var self = this;
    self.setCookie(name, '', -1);
  }

  getCookies (cname) {
    var self = this,
      data = {};
    document.cookie
      .split(';')
      .map(val => val.trim())
      .filter(val => !!val)
      .map(val => val.split('='))
      .map(parts => data[decodeURIComponent(parts[0])] = decodeURIComponent(parts[1]));
    return data;
  }

}

export default new Local();
